funsor/gaussian.py

```diff
@@ -155,21 +171,39 @@ def as_tensor(self):
         for i in rows:
             for j in cols:
                 if j not in self.parts[i]:
-                    shape = self.shape[:-2] + (i[1] - i[0], j[1] - j[0])
+                    shape = (i[1] - i[0], j[1] - j[0])
                     self.parts[i][j] = ops.new_zeros(prototype, shape)
 
         # Concatenate parts.
         # TODO This could be optimized into a single .reshape().cat().reshape() if
         #   all inputs are contiguous, thereby saving a memcopy.
-        columns = {i: ops.cat(-1, *[v for j, v in sorted(part.items())])
+        columns = {i: ops.cat(-1, *[ops.expand(v, self.shape[:-2] + v.shape[-2:]) for j, v in sorted(part.items())])
                    for i, part in self.parts.items()}
         result = ops.cat(-2, *[v for i, v in sorted(columns.items())])
+
         if not get_tracing_state():
             assert result.shape == self.shape
         return result
 
-
-def align_gaussian(new_inputs, old):
+    def __add__(self, other):
+        assert isinstance(other, BlockMatrix)
+        shape = broadcast_shape(self.shape, other.shape)
+        matrix = BlockMatrix(shape)
+        for part in set(self.parts.keys()) | set(other.parts.keys()):
+            a = self.parts[part]
```

Comment:
Won't this error if `part in other.parts` but `part not in self.parts`? And conversely for the following line?